services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedRegistrationService.java

```diff
@@ -0,0 +1,369 @@
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.MongoClientDeleteResult;
+import io.vertx.ext.mongo.MongoClientUpdateResult;
+
+/**
+ * This is an implementation of the device registration service and the device management service where data 
+ * is stored in a mongodb database.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/device-registration/">Device Registration API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+@Component
+@Qualifier("serviceImpl")
+@ConditionalOnProperty(name = "hono.app.type", havingValue = "mongodb", matchIfMissing = true)
+public class MongoDbBasedRegistrationService extends AbstractVerticle
+        implements DeviceManagementService, RegistrationService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedRegistrationService.class);
+    private MongoClient mongoClient;
+    private MongoDbBasedRegistrationConfigProperties config;
+    private MongoDbCallExecutor mongoDbCallExecutor;
+
+    /**
+     * Registration service, based on {@link AbstractRegistrationService}.
+     * <p>
+     * This helps work around Java's inability to inherit from multiple base classes. We create a new Registration
+     * service, overriding the implementation of {@link AbstractRegistrationService} with the implementation of our
+     * {@link MongoDbBasedRegistrationService#getDevice(String, String, Span)}.
+     */
+    private final AbstractRegistrationService registrationService = new AbstractRegistrationService() {
+
+        @Override
+        public Future<RegistrationResult> getDevice(final String tenantId, final String deviceId, final Span span) {
+            return MongoDbBasedRegistrationService.this.getDevice(tenantId, deviceId, span);
+        }
+
+        @Override
+        public Future<JsonArray> resolveGroupMembers(final String tenantId, final JsonArray viaGroups, final Span span) {
+            return MongoDbBasedRegistrationService.this.resolveGroupMembers(tenantId, viaGroups);
+        }
+    };
+
+    /**
+     * Creates an instance of the {@link MongoDbCallExecutor}.
+     *
+     * @param mongoDbCallExecutor An instance of the mongoDbCallExecutor.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = mongoDbCallExecutor;
+        this.mongoClient = this.mongoDbCallExecutor.getMongoClient();
+    }
+
+    @Autowired
+    public void setConfig(final MongoDbBasedRegistrationConfigProperties config) {
+        this.config = config;
+    }
+
+    public MongoDbBasedRegistrationConfigProperties getConfig() {
+        return config;
+    }
+
+    @Override
+    public void start(final Promise<Void> startPromise) {
+
+        mongoDbCallExecutor.createCollectionIndex(getConfig().getCollectionName(),
```

Comment:
if this is what I think it is ((re-)creating indices), then I believe it should be run periodically and independently from the start-up of this verticle

Reply:
@sophokles73 I understood from the above code, that they try creating indices during the startup of the component. If it doesn't succeed even after retries, then the component also fails to start up. 

> independently from the start-up of this verticle

Do you mean that even though the index creation fails, the component should start up?

>  then I believe it should be run periodically

Do you mean it should periodically try to create indexes until it is successfully created?

Reply:
I am not familiar with how indices are being created in maintained in Mongo DB. However, in general with DBMS, indices need to be created as part of the logical data model, e.g. when creating a table in a relational system. I wonder if this is a one time activity only or if Mongo requires the indices to be *refreshed* periodically. If so, I think this should be done e.g. from a side-car container cron job ...

Reply:
@sophokles73 As I understood from the _MongoDB_ documentation and what I have tested locally,  the indices once created remains there. Also _MongoDB_ does not recreate an index, if the index already exists for the same keys. I couldn't think of any place where we could run the index creation only once in our setup. Hence the start-up section of the component seems to be the next appropriate place for me. 

Also what I understood from the _MongoDB_ docs is that the index creation process holds an exclusive lock on the specified collection at the start and end of the index build. All subsequent operations on the collection must wait until the exclusive lock is released. But it allows interleaving read and write operations during the _majority_ of the index build. This tells me that its slightly better to run it during the start-up of the component rather independently from the start-up.

Reply:
@sophokles7 Would you mind taking a look at the above comment? WDYT about it?

Reply:
Then let's keep it like it is ...